### PR TITLE
feat(pallets/subspace): add arbitrary uri parameter

### PR DESCRIPTION
**Pull request by aripiprazole**

### CODE REVIEW

**Comment by steinerkelvin** [comment]
When I said "arbitrary URI" it was just a description of what would go in the field 😅

The field should be named `metadata` or `metadata_uri` or something like that. @saiintbrisson

File: pallets/subspace/src/lib.rs [hunk 1]
```diff
@@ -738,6 +739,14 @@ pub mod pallet {
...
         StorageDoubleMap<_, Twox64Concat, u16, Twox64Concat, u16, Vec<u8>, ValueQuery>;
 
+    #[pallet::type_value]
+    pub fn DefaultArbitraryUri<T: Config>() -> Vec<u8> {
+        vec![]
+    }
+    #[pallet::storage] // --- DMAP ( netuid, uid ) --> module_address
+    pub type ArbitraryUri<T: Config> =
+        StorageDoubleMap<_, Twox64Concat, u16, Twox64Concat, u16, Vec<u8>, ValueQuery>;
+
```
Comment (saiintbrisson) [inline pallets/subspace/src/lib.rs:749]:
```suggestion
    #[pallet::storage] // --- DMAP ( netuid, uid ) --> module_address
    pub type ArbitraryUri<T: Config> =
        StorageDoubleMap<_, Twox64Concat, u16, Twox64Concat, u16, Vec<u8>>;

```

File: pallets/subspace/src/lib.rs [hunk 2]
```diff
@@ -349,6 +349,7 @@ pub mod pallet {
         pub name: Vec<u8>,
         pub address: Vec<u8>,
         pub delegation_fee: Percent,
+        pub arbitrary_uri: Vec<u8>,
```
Comment (saiintbrisson) [inline pallets/subspace/src/lib.rs:352]:
```suggestion
        pub arbitrary_uri: Option<Vec<u8>>,
```

**Review by saiintbrisson — CHANGES_REQUESTED** [review]
Rename it as kelvin said and you have to add a way to change the metadata too. Look into how `ModuleChangeset` works

**Comment by steinerkelvin** [comment]
Also, this field should have a limited size. Max 256 bytes or so, so people don't dump Shakespeare inside the module metadata. Just enough for a secure hash/id/uri of the content.


According to the oracle (Claude):

> The maximum size of an IPFS CID (Content Identifier) string depends on the CID version and the multibase encoding used. Here are the details:
> 
> CID Version 0 (CIDv0):
> 
> CIDv0 uses base58btc encoding.
> The maximum length of a CIDv0 string is 46 characters.
> Example: QmYwAPJzv5CZsnA625s3Xf2nemtYgPpHdWEz79ojWnPbdG
> CID Version 1 (CIDv1):
> 
> CIDv1 supports various multibase encodings, but the most commonly used are base32 and base58btc.
> The maximum length of a CIDv1 string depends on the multibase encoding:
> base32: The maximum length is 62 characters.
> Example: bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq
> base58btc: The maximum length is 59 characters.
> Example: zdj7WWeQ43G6JJvLWQWZpyHuAMq6uYWRjkBXFad11vE2LHhQ7
> It's important to note that the actual length of a CID string can vary depending on the content being addressed and the specific multicodec and multihash used within the CID.
> 
> In summary, the maximum size of an IPFS CID string is:
> 
> CIDv0 (base58btc): 46 characters
> CIDv1 (base32): 62 characters
> CIDv1 (base58btc): 59 characters

https://github.com/multiformats/cid#decoding-algorithm

https://cid.ipfs.tech/#bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq

**Comment by aripiprazole** [comment]
> Rename it as kelvin said and you have to add a way to change the metadata too. Look into how `ModuleChangeset` works

Should I add it into changeset struct?

**Comment by aripiprazole** [comment]
> Also, this field should have a limited size. Max 256 bytes or so, so people don't dump Shakespeare inside the module metadata. Just enough for a secure hash/id/uri of the content.
> 
> According to the oracle (Claude):
> 
> > The maximum size of an IPFS CID (Content Identifier) string depends on the CID version and the multibase encoding used. Here are the details:
> > CID Version 0 (CIDv0):
> > CIDv0 uses base58btc encoding.
> > The maximum length of a CIDv0 string is 46 characters.
> > Example: QmYwAPJzv5CZsnA625s3Xf2nemtYgPpHdWEz79ojWnPbdG
> > CID Version 1 (CIDv1):
> > CIDv1 supports various multibase encodings, but the most commonly used are base32 and base58btc.
> > The maximum length of a CIDv1 string depends on the multibase encoding:
> > base32: The maximum length is 62 characters.
> > Example: bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq
> > base58btc: The maximum length is 59 characters.
> > Example: zdj7WWeQ43G6JJvLWQWZpyHuAMq6uYWRjkBXFad11vE2LHhQ7
> > It's important to note that the actual length of a CID string can vary depending on the content being addressed and the specific multicodec and multihash used within the CID.
> > In summary, the maximum size of an IPFS CID string is:
> > CIDv0 (base58btc): 46 characters
> > CIDv1 (base32): 62 characters
> > CIDv1 (base58btc): 59 characters
> 
> https://github.com/multiformats/cid#decoding-algorithm
> 
> https://cid.ipfs.tech/#bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq

@steinerkelvin Where do I put the validation here? Is it only when we are trying to update (aka ModuleChangeset) or there's another place to add it?

**Comment by steinerkelvin** [comment]
> @steinerkelvin Where do I put the validation here? Is it only when we are trying to update (aka ModuleChangeset) or there's another place to add it?

Oh, this was just a discussing size. The only validation subspace-side should be that it's a valid UTF8 string, and the size limit.

**Comment by saiintbrisson** [comment]
> Where do I put the validation here?
 
On the apply function of ModuleChangeset

**Pull request merged**